jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContentProducer.java

```diff
@@ -316,119 +316,144 @@ public boolean isReady()
-                            _rawContent = refreshedRawContent;
-                        _error = _rawContent.getError() != null;
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("refreshed raw content: {} {}", _rawContent, this);
-                    }
+            if (LOG.isDebugEnabled())
+                LOG.debug("transforming raw content {}", this);
+            transformRawContent();
+        }
+    }
 
-                    _transformedContent = _rawContent;
-                    break;
-                }
-                else if (_rawContent.isEmpty())
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("nulling depleted raw content {}", this);
-                    _rawContent.succeeded();
-                    _rawContent = null;
-                    _rawContent = produceRawContent();
-                    if (_rawContent == null)
-                    {
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("produced null raw content, returning null, {}", this);
-                        break;
-                    }
-                }
-                else
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("raw content is not empty {}", this);
-                }
+    private void transformRawContent()
+    {
+        if (_interceptor != null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("intercepting raw content {}", this);
+            _transformedContent = intercept();
+
+            // If the interceptor generated a special content, _rawContent must become that special content.
+            if (_transformedContent != null && _transformedContent.isSpecial() && _transformedContent != _rawContent)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("interceptor generated a special content, _rawContent must become that special content {}", this);
+                _rawContent.succeeded();
+                _rawContent = _transformedContent;
+                return;
             }
-            else
+
+            // If the interceptor generated a null content, recycle the raw content now if its empty.
+            if (_transformedContent == null && _rawContent.isEmpty() && !_rawContent.isSpecial())
             {
                 if (LOG.isDebugEnabled())
-                    LOG.debug("transformed content is not empty {}", this);
+                    LOG.debug("interceptor generated a null content, recycling the empty raw content now {}", this);
+                _rawContent.succeeded();
+                _rawContent = null;
+                return;
+            }
+
+            // If the interceptor returned the raw content, recycle the raw content now if its empty.
+            if (_transformedContent == _rawContent && _rawContent.isEmpty() && !_rawContent.isSpecial())
```

Comment:
This `if` condition basically states that interceptors MUST completely consume the content.

It is what it is, but A) needs to be javadoc'ed and B) needs a safeguard (and a test case) because otherwise the implementation goes into a spin loop!

Reply:
Good catch, let me add the necessary javadoc, safeguard check and tests.